### PR TITLE
added ability to pass in validation as an optional paramater

### PR DESCRIPTION
**Pull request by validkeys**
Thanks for creating this lib!

I wanted to be able to run validations on my virtual fields but was unable to do so given the original setup. This PR adds the ability to pass in validation as an optional second parameter to the "field" method. If a validation object has been passed in, the set method calls the ```this.setDataValue``` method in order to trigger validation.

Let me know if you have any feedback :)

### CODE REVIEW

**Comment by validkeys** [comment]
There doesn't appear to be a travis.yml file in the master branch off of which i based this PR.

File: index.js [hunk 1]
```diff
@@ -47,12 +47,23 @@ EncryptedField.prototype.vault = function(name) {
...
+    config              = config || {};
+    var hasValidations  = [undefined,null].indexOf(typeof config.validate) === -1;
+
+    config.validate  = config.validate || {};
+
+    var fieldType = (config.hasOwnProperty('type')) ? self.Sequelize.VIRTUAL(config.type) : self.Sequelize.VIRTUAL;
 
     return {
-        type: self.Sequelize.VIRTUAL,
+        type: fieldType,
```
Comment (defunctzombie) [inline index.js]: Why wouldn't the field be virtual? We don't want the field itself to be saved as it is just a proxy for the decrypted values from the vault. I am concerned that by not having it be virtual, sequelize will try to load the field.

**Comment by defunctzombie** [comment]
Please fix your indentation. You used 2 spaces where all of the existing indentation is 4.

File: index.js [hunk 2]
```diff
@@ -47,12 +47,20 @@ EncryptedField.prototype.vault = function(name) {
     }
 };
 
-EncryptedField.prototype.field = function(name) {
-    var self = this;
+EncryptedField.prototype.field = function(name, validation) {
+    var self            = this;
```
Comment (validkeys) [inline index.js]: sorry -- my OCD got the better of me. Will revert.

File: index.js [hunk 3]
```diff
@@ -47,12 +47,20 @@ EncryptedField.prototype.vault = function(name) {
...
+    var hasValidations  = [undefined,null].indexOf(typeof validation) === -1;
+
+    validation  = validation || {};
 
     return {
         type: self.Sequelize.VIRTUAL,
         set: function set_encrypted(val) {
+
+            if (hasValidations) {
+              this.setDataValue(name, val);
```
Comment (validkeys) [inline index.js]: whoops. will fix. thanks!

File: index.js (hunk 1 above)
Comment (validkeys) [inline index.js]: It's not that the field wouldn't be virtual. Virtual fields can optionally take a return type (http://docs.sequelizejs.com/en/latest/api/datatypes/#virtual)

So a generic virtual field would be:

```
Sequelize.VIRTUAL
```

But a virtual field that should return a boolean type would be:

```
Sequelize.VIRTUAL(Sequelize.BOOLEAN...
```

So the actual type will always be virtual, I'm just doing a check to see if the user explicitly included a return type. I will clean up my code and rebase off of master to include the last PR acceptance.

**Comment by validkeys** [comment]
@defunctzombie so I cleaned up the code and rebased off of master. So to recap, the main additions here are that you get all of the same configuration options are you would when creating a regular sequelize field:

```javascript
    private_2: enc_fields.field('private_2', {
      type: Sequelize.TEXT,
      validate: {
        isInt: true  
      },
      allowNull: false,
      defaultValue: null
    })
```

If you pass in the "type", your field will be cast as that type using the built-in functionality of sequelize. You can pass in validation rules just like you would a regular field and you can also supply a default value. One of the issues I was having was when creating models where the field is not in the payload, the encrypted field was showing up like this:

```
{"encrypted": {"someField":"undefined"}}
```

So passing in the the default value, the getter on the virtual field returns your defaultValue if the value in the encrypted field is empty.

Finally, the ability to pass in the allowNull attr as you would a regular property.

I've also added tests for each of the above.

**Comment by defunctzombie** [comment]
@validkeys awesome explanation. I understand now and this is a great improvement. Could you please squash the commits down to one right now the diff seems wrong to me (somehow one of the commits from master made it into this PR?)

**Comment by validkeys** [comment]
@defunctzombie something happened after I rebased. It included files added from your master commits. My git abilities are somewhat lacking -- not sure what to do at this point. If you've got any ideas, let me know. Otherwise, I'll search around to see if I can find what's happening.

**Comment by validkeys** [comment]
@defunctzombie it looks like it might just be github's view here that's showing the files that were changed. The actual pull request only has the 3 changed files.

screen shot 2016-10-01 at 4 41 05 pm

**Comment by validkeys** [comment]
@defunctzombie I just keep digging myself deeper. I am going to close this PR and submit a new one. Sorry about this. GIT has always been my weakness :)

**Pull request closed without merging**